File: java/midp/com/nokia/mid/ui/frameanimator/FrameAnimator.java
package com.nokia.mid.ui.frameanimator;

public class FrameAnimator {

    public static final int FRAME_ANIMATOR_MAX_CONCURRENT = 5;
    public static final int FRAME_ANIMATOR_VERTICAL = 0;
    public static final int FRAME_ANIMATOR_HORIZONTAL = 1;
    public static final int FRAME_ANIMATOR_FREE_ANGLE = 2;
    public static final int FRAME_ANIMATOR_FRICTION_LOW = 0;
    public static final int FRAME_ANIMATOR_FRICTION_MEDIUM = 1;
    public static final int FRAME_ANIMATOR_FRICTION_HIGH = 2;

    protected int actionType;
    protected int actionID;

    protected static int _numRegistered;

    public FrameAnimator() {
        System.out.println("warning: FrameAnimator() not implemented");
    }

    public boolean register(int x, int y, short maxFps, short maxPps, FrameAnimatorListener listener) {
        System.out.println("warning: FrameAnimator.register(IISSL...FrameAnimatorListener;)Z not implemented (" +
                           x + ", " + y + ", " + maxFps + ", " + maxPps + ", " + listener + ")");
        _numRegistered++;
        return true;
    }

    public void unregister() {
        System.out.println("warning: FrameAnimator.unregister()V not implemented");
        _numRegistered--;
    }

    public void drag(int x, int y) {
        System.out.println("warning: FrameAnimator.drag(II)V not implemented (" + x + ", " + y + ")");
    }

    public void kineticScroll(int speed, int direction, int friction, float angle) {
        System.out.println("warning: FrameAnimator.kineticScroll(IIIF)V not implemented (" +
                           speed + ", " + direction + ", " + friction + ", " + angle + ")");
    }

    public void limitedKineticScroll(int speed, int direction, int friction, float angle, int limitUp, int limitDown) {
        System.out.println("warning: FrameAnimator.limitedKineticScroll(IIIFII)V not implemented (" +
                           speed + ", " + direction + ", " + friction + ", " + angle + ", " + limitUp + ", " +
                           limitDown + ")");
    }

    public void stop() {
        System.out.println("warning: FrameAnimator.stop()V not implemented");
    }

    public boolean isRegistered() {
        System.out.println("warning: FrameAnimator.isRegistered()Z not implemented");
        return true;
    }

    public static int getNumRegisteredFrameAnimators() {
        System.out.println("warning: FrameAnimator.getNumRegisteredFrameAnimators()I not implemented");
        return _numRegistered;
    }

}

File: midp/frameanimator.js
/* -*- Mode: Java; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
/* vim: set shiftwidth=2 tabstop=2 autoindent cindent expandtab: */

'use strict';

var FrameAnimator = function() {};

FrameAnimator.numRegistered = 0;

FrameAnimator.prototype._isRegistered = false;

FrameAnimator.prototype.register = function(x, y, maxFps, maxPps, listener) {
  this.x = x;
  this.y = y;
  this.maxFps = maxFps;
  this.maxPps = maxPps;
  this.listener = listener;
  this._isRegistered = true;
  ++FrameAnimator.numRegistered;
};

FrameAnimator.prototype.unregister = function() {
  this.x = null;
  this.y = null;
  this.maxFps = null;
  this.maxPps = null;
  this.listener = null;
  this._isRegistered = false;
  --FrameAnimator.numRegistered;
};

FrameAnimator.prototype.isRegistered = function() {
  return this._isRegistered;
};

Native["com/nokia/mid/ui/frameanimator/FrameAnimator.init.()V"] = function() {
  this.nativeObject = new FrameAnimator();
};

Native["com/nokia/mid/ui/frameanimator/FrameAnimator.register.(IISSLcom/nokia/mid/ui/frameanimator/FrameAnimatorListener;)Z"] = function(x, y, maxFps, maxPps, listener) {
  if (this.nativeObject.isRegistered()) {
    throw $.newIllegalStateException("FrameAnimator already registered");
  }

  if (!listener) {
    throw $.newNullPointerException("listener is null");
  }

  if (x < -65535 || x > 65535 || y < -65535 || y > 65535) {
    throw $.newIllegalArgumentException("coordinate out of bounds");
  }

  // XXX return false if FrameAnimator.numRegistered >= FRAME_ANIMATOR_MAX_CONCURRENT

  this.nativeObject.register(x, y, maxFps, maxPps, listener);
  return 1;
};

Native["com/nokia/mid/ui/frameanimator/FrameAnimator.unregister.()V"] = function() {
  if (!this.nativeObject.isRegistered()) {
    throw $.newIllegalStateException("FrameAnimator not registered");
  }

  this.nativeObject.unregister();
};

addUnimplementedNative("com/nokia/mid/ui/frameanimator/FrameAnimator.drag.(II)V");
addUnimplementedNative("com/nokia/mid/ui/frameanimator/FrameAnimator.kineticScroll.(IIIF)V");
addUnimplementedNative("com/nokia/mid/ui/frameanimator/FrameAnimator.limitedKineticScroll.(IIIFII)V");
addUnimplementedNative("com/nokia/mid/ui/frameanimator/FrameAnimator.stop.()V");

Native["com/nokia/mid/ui/frameanimator/FrameAnimator.isRegistered.()Z"] = function() {
  return this.nativeObject.isRegistered() ? 1 : 0;
};

Native["com/nokia/mid/ui/frameanimator/FrameAnimator.getNumRegisteredFrameAnimators.()I"] = function() {
  return FrameAnimator.numRegistered;
};
